File: Advance-Data-Structure/queue.cpp
#include <bits/stdc++.h>
using namespace std;
#define SIZE 10

template <class Type>
class Queue {

private:
    Type *arr;
    int head;
    int tail;
    int capacity;

public:
    Queue(int size) {
        arr = new Type[size];
        head = 0;
        tail = 0;
        capacity = size;
    }

    void enqueue(Type item) {
        if (isFull()) {
            printf("Error, stack is full\n");
            exit(EXIT_FAILURE);
        }
        arr[tail] = item;
        tail = (tail+1) % capacity;
    }

    Type dequeue() {
        if (isEmpty()) {
            printf("Error, stack is empty\n");
            exit(EXIT_FAILURE);
        }
        Type item = arr[head];
        head = (head + 1) % capacity;
        return item;
    }

    bool isFull() {
        return ((tail + 1) % capacity == head);
    }

    bool isEmpty() {
        return head == tail;
    }
};

int main() {
    Queue <string> q(5);
    
    q.enqueue("brac");
    q.enqueue("aiub");
    q.enqueue("du");
    
    cout << q.dequeue() << endl;
    cout << q.dequeue() << endl;
    cout << q.dequeue() << endl;
    cout << q.dequeue() << endl;
}

File: Advance-Data-Structure/PYTHON/queue.py
from collections import deque

class Queue:
    def __init__(self):
        self.items = deque()
        
    def is_empty(self):
        return not self.items
    
    def enqueue(self, item):
        self.items.append(item)
    
    def dequeue(self):
        if self.is_empty():
            print('Queue empty...')
            raise SystemExit
        return self.items.popleft()
    
    def size(self):
        return len(self.items)
    
    def peek(self):
        if self.is_empty():
            print('Queue empty...')
            raise SystemExit
        return self.items[0]
    
    def __str__(self):
        return str(self.items)
    
if __name__ == '__main__':
    q = Queue()
    print(q)
    print(q.is_empty())
    q.enqueue(1)
    q.enqueue(2)
    print(q)
    print(q.dequeue())
    print(q.peek())
    print(q)
    print(q.dequeue())
    print(q.dequeue())